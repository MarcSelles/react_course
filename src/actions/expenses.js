import database from '../firebase/firebase';

// Component calls action generator
// Action generator returns function
// Component dispatches function
// Function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
        const expense = { description, note, amount, createdAt }

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {} ) => ({
    type: 'REMOVE_EXPENSE',
    id
  });

export const startRemoveExpense = (expenseID) => {
    return (dispatch) => {
        return database.ref(`expenses/${expenseID}`)
            .remove()
            .then((snapshot) => {
                dispatch(removeExpense({id: expenseID}));
            }, (e) => console.log(e));
    };
};

// EDIT_EXPENSE
export const editExpense = ( id, updates ) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`)
            .update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            }, (e) => console.log(e));
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                const databaseExpenses = [];
                snapshot.forEach((childSnapshot) => {
                    databaseExpenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(databaseExpenses));
            })
        };
};