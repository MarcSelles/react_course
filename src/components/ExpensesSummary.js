import React from 'react';
import { connect } from 'react-redux';
import expensesTotal from '../selector/expenses-total';
import selectExpenses from '../selector/expenses';
import numeral from 'numeral';
let locales = require('numeral/locales');
numeral.locale('nl-nl');

export const ExpensesSummary = (props) => (
    <div>
        <h2>
            Viewing {props.expenses.length} expense{props.expenses.length === 1 ? '' : 's'} totalling {numeral(expensesTotal(props.expenses)/100).format('$0,0.00')}
        </h2>
    </div>
);

const mapStateToProps = (state) => ({ expenses: selectExpenses(state.expenses, state.filters) });

export default connect(mapStateToProps)(ExpensesSummary);