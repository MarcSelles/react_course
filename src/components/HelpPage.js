import React from 'react';

const HelpPage = () => (
    <div>Dit is Help</div>
);

export default HelpPage;