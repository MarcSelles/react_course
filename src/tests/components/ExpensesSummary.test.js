import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test ('should render the summary with items correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test ('should render the summary with one item correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]}/>);
    expect(wrapper).toMatchSnapshot();
});

test ('should render the summary without items correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});