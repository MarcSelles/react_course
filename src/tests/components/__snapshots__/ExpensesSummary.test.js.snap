// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the summary with items correctly 1`] = `
<div>
  <h2>
    Viewing 
    3
     expense
    s
     totalling 
    € 1.141,95
  </h2>
</div>
`;

exports[`should render the summary with one item correctly 1`] = `
<div>
  <h2>
    Viewing 
    1
     expense
     totalling 
    € 1,95
  </h2>
</div>
`;

exports[`should render the summary without items correctly 1`] = `
<div>
  <h2>
    Viewing 
    0
     expense
    s
     totalling 
    € 0,00
  </h2>
</div>
`;
